File: 0010-regular-expression-matching/0010-regular-expression-matching.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        @cache
        def dfs(i, j):
            # If pattern is fully traversed, return True if the string is also fully traversed
            if j >= n:
                return i == m
            
            # If the next character in pattern is '*', handle zero or more occurrences
            if j + 1 < n and p[j + 1] == '*':
                return dfs(i, j + 2) or (
                    i < m and (s[i] == p[j] or p[j] == '.') and dfs(i + 1, j)
                )
            
            # Match current characters and move to the next
            return i < m and (s[i] == p[j] or p[j] == '.') and dfs(i + 1, j + 1)

        m, n = len(s), len(p)
        return dfs(0, 0)

File: 0010-regular-expression-matching/0010-regular-expression-matching.java
class Solution {
    private Boolean[][] f; // Memoization table to store results of subproblems
    private String s;
    private String p;
    private int m; // Length of string s
    private int n; // Length of pattern p

    public boolean isMatch(String s, String p) {
        m = s.length();
        n = p.length();
        f = new Boolean[m + 1][n + 1]; // Initialize memoization table
        this.s = s;
        this.p = p;
        return dfs(0, 0); // Start DFS from the beginning of both string and pattern
    }

    private boolean dfs(int i, int j) {
        if (j >= n) {
            return i == m; // If pattern is exhausted, check if string is also exhausted
        }
        if (f[i][j] != null) {
            return f[i][j]; // Return memoized result if already computed
        }
        boolean res = false;
        if (j + 1 < n && p.charAt(j + 1) == '*') {
            // Two possibilities: ignore the '*' or use it to match characters
            res = dfs(i, j + 2) 
                || (i < m && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.') && dfs(i + 1, j));
        } else {
            // Regular character match or '.' wildcard
            res = i < m && (s.charAt(i) == p.charAt(j) || p.charAt(j) == '.') && dfs(i + 1, j + 1);
        }
        return f[i][j] = res; // Store the result in the memoization table
    }
}

